Explicit Padding in CBuffers Proposal (#311)
llvm/wg-hlsl: proposals/NNNN-explicit-padding.md

> <!-- {% raw %} -->
> 
> # Explicit Padding in Structs and CBuffer Arrays
> 
> * Proposal: [NNNN](NNNN-explicit-padding.md)
> * Author(s): [Justin Bogner](https://github.com/bogner)
> * Status: **Design In Progress**

Reviewer: ```suggestion
---
title: "[NNNN] - Explicit Padding in Structs and CBuffer Arrays"
params:
    status: Design In Progress
    authors:
        - bogner: Justin Bogner
---

```

## Introduction

> We introduce an explicit padding type for HLSL, and construct cbuffer arrays
> and structs that are annotated with `packoffset` or `vk::offset` using this
> type to unambiguously lay out these objects.

Reviewer 2: ```suggestion
We introduce an explicit padding type for HLSL, and construct cbuffer arrays,
structs, and cbuffer elements that are annotated with `packoffset` or `vk::offset` using this
type to unambiguously lay out these objects.
```

Reviewer 2: Structs in cbuffer must also start on a 16-byte boundary and we need to insert padding before if needed: https://godbolt.org/z/YEjc7TEY9

- Arrays in CBuffers may have padding in between members, but crucially they do
  not have padding after the last member. This needs special handling to
  represent.

- There are two HLSL constructs that may introduce arbitrary padding to a
  struct. In a cbuffer, the `packoffset` attribute specifies the offset of a
  member, and outside of cbuffers, the `vk::offset` vulkan attribute may do the
  same.

- Simply padding structures with `i8` as is typical with ABI-related padding
  makes it difficult to recover which struct elements are padding vs which are
  subobjects. This matters in some backends, and is specifically important for
  SPIR-V where we need to map a logical indices into the struct into physical
  offsets.

## Proposed solution

Introduce an explicit padding type and use this type for the padding in the
various constructs that need it.

The padding type will be a target type with a parameter for the size in bytes.
For DirectX, this would look like `target("dx.Layout", 4)` for 4 bytes of
padding. For SPIR-V, `target("spirv.Padding", 8)` for 8 bytes of padding. We
may attempt to come up with a first-class type in LLVM for these purposes in
the future.

> ### Structs with annotations

Reviewer 2: I would probably move this section after "### Arrays in a cbuffer" because this is way less common than ararys in cbuffer.

lexicographical order:

```hlsl
cbuffer cb0 {
    int x : packoffset(c0.y);
    float y : packoffset(c0.x);
}
```

To support this, we need to create the underlying LLVM type in the order that
matches the packoffsets rather than the order as written, so we would end up
with `{ float, i32 }` here, losing the lexicographical order. This is probably
okay since we need to create artificial types for cbuffers anyway (such as when
we filter out resource types that are declared within the cbuffer), but may not
make for a particularly good debugging experience.

### Arrays in a cbuffer

CBuffers in HLSL have very specific layout rules. Each "object" starts at a
16-byte boundary, which is mostly explainable as a 16-byte alignment
requirement, but applies to array elements rather than the array itself in a
way that doesn't match the general language. We can emulate this with an array
of objects that consist of a struct containing the element type and padding to
16 bytes, followed by a single instance of the element type itself. See
[CBuffer Padded arrays at the HLSL-level] for details.

[CBuffer Padded arrays at the HLSL-level]: #cbuffer-padded-arrays-at-the-hlsl-level

> 

Reviewer 2: Add section "### Struct in a cbuffer" (here or before the array section) describing that structs in cbuffer must also start on a 16-byte boundary and we need to insert padding before if as needed.

Author: I found a separate section for this more confusing so I instead reworked this a bit to talk about scalars, vectors, structs, and arrays explicitly.

`cb0`:

```hlsl
cbuffer cb0 : register(b0) {
  int x : packoffset(c0.y);
  float y : packoffset(c1.z);
}
```

```llvm
%__cblayout_cb0 = type <{
  [4 x %pad8],
  i32,
  [16 x %pad8],
  float
}>
```

For arrays, we'll have padding within elements to fill to a 16-byte boundary,
and padding before arrays in order for them to start at 16-byte boundaries.
Consider `cb1`:

```hlsl
cbuffer cb1 : register(b0) {
  float a1[3];        // offset   0,  size 4  (+12) * 3
  double3 a2[2];      // offset   48, size 24  (+8) * 2
  uint4 a3[2];        // offset  112, size 16       * 2
  float16_t a4[2][2]; // offset  144, size  2 (+14) * 4
}
```

```llvm
%__cblayout_cb1 = type <{
  <{ [2 x <{ float, target("dx.Padding", 12) }>], float }>,
  target("dx.Padding", 12"),
  <{ [1 x <{ <3 x double>, target("dx.Padding", 8) }>], <3 x double> }>,
  target("dx.Padding", 8),
  [ 2 x <4 x i32> ],
  <{ [3 x <{ half, target("dx.Padding", 14) }>], half }>
}>
```

[__cblayout]: https://github.com/llvm/wg-hlsl/blob/4570a9cfc5c4b1e5bc0b773a6fb7b22014ac6d3b/proposals/0016-constant-buffers.md#lowering-constant-buffer-resources-to-llvm-ir "Lowering Constant Buffer Resources to LLVM IR"

> 

Reviewer 2: Add example of cbuffer with struct that has padding before.

clang. Here we can recognize that the source of the copy is in the cbuffer
address space and break the copy up into elementwise pieces.

## Alternatives considered

See [llvm-project/wg-hlsl#171] for the previous attempt at representing these
types.

Regarding the padding type, we considered the following options:

- A first class LLVM type called `pad8`, which is equivalent but distinct from
  `i8`. This would need an RFC to the wider LLVM community and would need to be
  useful in other contexts (such as ABI-mandated padding).
- A well-known named type `%pad8`, defined as a named struct containing a
  single `i8`. This is the simplest option but requires backends that are
  interested in this type to participate in a secret handshake.
- Target types such as `target("dx.pad8")` and `target("spirv.pad8")`. This is
  somewhat awkward because the type isn't really tied to a target, but target
  types need to be. Targets that don't need to differentiate between padding
  and actual members could simply use `i8`.
  
We settled on a target type with a size parameter for now.

[llvm-project/wg-hlsl#171]: https://github.com/llvm/wg-hlsl/pull/171

> <!-- {% endraw %} -->

Reviewer: ```suggestion
```

  subobjects. This matters in some backends, and is specifically important for
  SPIR-V where we need to map a logical indices into the struct into physical
  offsets.

## Proposed solution

Introduce an explicit padding type and use this type for the padding in the
various constructs that need it.

The padding type will be defined as one of the following:

- A first class LLVM type called `pad8`, which is equivalent but distinct from
  `i8`. This would need an RFC to the wider LLVM community and would need to be
  useful in other contexts (such as ABI-mandated padding).
- A well-known named type `%pad8`, defined as a named struct containing a
  single `i8`. This is the simplest option but requires backends that are
  interested in this type to participate in a secret handshake.
- Target types such as `target("dx.pad8")` and `target("spirv.pad8")`. This is
  somewhat awkward because the type isn't really tied to a target, but target
  types need to be. Targets that don't need to differentiate between padding
  and actual members could simply use `i8`.

> TODO: Choose one of these three options and move the others to the

> > "alternatives" section.

Reviewer 3: 1. If you use the target type, you can do something like we do for the `vk::SpirvType` where the size is a parameter. So the padding is always just one instance of the type, not an array. However, not a significant advantage.
2. The "well known name" solution has the problem that theoretically, the optimizer does not know the name is special, and might change it in some way. Let me know if there is something in the llvm-ir spec that would guarantee that it will not be changed.
3. The `pad8` is a great idea if it is accepted by the llvm community. But it could take a while.

My thoughts are to do the target type for now. See if the LLVM community is interested in `pad8`. If so, we can switch to it when it is added.

I do not have strong opinions on this, and I will not hold up this proposal if you do something different.

}>
```

For structs, we add padding to align them as appropriate:

```hlsl
struct S {
  int v;
};
cbuffer cb1 : register(b0) {
  int i; // offset   0,  size 4  (+12)
  S s;   // offset  16,  size 4
  int j; // offset  20,  size 4
}

```llvm
%__cblayout_cb1 = type <{
  i32, target("dx.Padding", 12),

>   i32,

Reviewer 2: Are we flattening the struct `S` here? I though we preserve it.

Author: The struct is preserved. I've fixed the example.
